### PR TITLE
Handle edge cases for list of pending acknowledgements

### PR DESCRIPTION
**Pull request by aggarw13**
Handle the following 2 edge cases for entries in the list of pending acknowledgments (or `MQTTAgentContext_t.pPendingAcks`): 
1. When an MQTT session is resumed, but the previous connection of the session had un-acknowledged SUBSCRIBE and UNSUBSCIBE requests, the list needs to be cleared of entries for those operations. This PR updates the `MQTT_ResumeSession` API for the clean-up operation.
2. In the unlikely case when an operation is unkacknowledged, and there occur 65535 operations after it that cause the packet ID value to wrap around, then the 65536th operation, if needing acknowledgement, will have the same packet ID as the still packet ID still pending acknowledgment in the list (packet ID is represented as a 16 bit integer). This PR handles this case by returning an `MQTTStateCollision` error in that case.

### CODE REVIEW

File: source/mqtt_agent.c [hunk 1]
```diff
@@ -295,35 +305,65 @@ static bool isSpaceInPendingAckList( const MQTTAgentContext_t * pAgentContext )
...
+             * the list.
+             * Note: This is an unlikely edge case which represents that a packet ID
+             * didn't receive acknowledgement, but subsequent SUBSCRIBE/PUBLISH operations
+             * representing 65535 packet IDs were successful that caused the bit packet
+             * ID value to wrap around and reached the same packet ID as that was still
+             * pending acknowledgement.
+             */
+            status = MQTTStateCollision;
+            LogError( ( "Failed to add operation to list of pending acknowledgements: Existing entry found :"
+                        "for same packet: PacketId=%u\n", packetId ) );
```
Comment (muneebahmed10) [inline source/mqtt_agent.c:348]: We've been dropping the 'e' everywhere else in this repo. to match the most common American spelling.
```suggestion
            /* Check whether there exists a duplicate entry for pending
             * acknowledgment for the same packet ID that we want to add to
             * the list.
             * Note: This is an unlikely edge case which represents that a packet ID
             * didn't receive acknowledgment, but subsequent SUBSCRIBE/PUBLISH operations
             * representing 65535 packet IDs were successful that caused the bit packet
             * ID value to wrap around and reached the same packet ID as that was still
             * pending acknowledgment.
             */
            status = MQTTStateCollision;
            LogError( ( "Failed to add operation to list of pending acknowledgments: Existing entry found :"
                        "for same packet: PacketId=%u\n", packetId ) );
```

File: source/mqtt_agent.c [hunk 2]
```diff
@@ -295,35 +305,65 @@ static bool isSpaceInPendingAckList( const MQTTAgentContext_t * pAgentContext )
...
+        if( ( unusedPos == MQTT_AGENT_MAX_OUTSTANDING_ACKS ) &&
+            ( pendingAcks[ i ].packetId == MQTT_PACKET_ID_INVALID ) )
         {
-            pendingAcks[ i ].packetId = packetId;
-            pendingAcks[ i ].pOriginalCommand = pCommand;
-            ackAdded = true;
+            unusedPos = i;
+            status = MQTTSuccess;
+        }
+        else if( pendingAcks[ i ].packetId == packetId )
```
Comment (muneebahmed10) [inline source/mqtt_agent.c:335]: Please either add an empty `else` block for MISRA, or change this `else if` to an `if`

File: source/mqtt_agent.c [hunk 3]
```diff
@@ -295,35 +305,65 @@ static bool isSpaceInPendingAckList( const MQTTAgentContext_t * pAgentContext )
...
+    if( status == MQTTSuccess )
+    {
+        pendingAcks[ unusedPos ].packetId = packetId;
+        pendingAcks[ unusedPos ].pOriginalCommand = pCommand;
+    }
+    else if( status == MQTTNoMemory )
+    {
+        LogError( ( "Failed to add operation to list of pending acknowledgements: No memory available:"
+                    "PacketId=%u\n", packetId ) );
+    }
```
Comment (muneebahmed10) [inline source/mqtt_agent.c:364]:
```suggestion
        LogError( ( "Failed to add operation to list of pending acknowledgments: No memory available:"
                    "PacketId=%u\n", packetId ) );
    }
    else
    {
        /* Empty else MISRA 15.7 */
    }
```

File: source/mqtt_agent.c [hunk 4]
```diff
@@ -515,18 +555,8 @@ static MQTTStatus_t processCommand( MQTTAgentContext_t * pMqttAgentContext,
...
-        if( !ackAdded )
-        {
-            LogError( ( "No memory to wait for acknowledgment for packet %u\n", commandOutParams.packetId ) );
-
-            /* All operations that can wait for acks (publish, subscribe,
-             * unsubscribe) require a context. */
-            operationStatus = MQTTNoMemory;
-        }
+        operationStatus = addAwaitingOperation( pMqttAgentContext, commandOutParams.packetId, pCommand );
+        ackAdded = ( operationStatus == MQTTSuccess ) ? true : false;
```
Comment (muneebahmed10) [inline source/mqtt_agent.c:559]: The last part isn't really necessary
```suggestion
        ackAdded = ( operationStatus == MQTTSuccess );
```

File: test/unit-test/mqtt_agent_utest.c [hunk 5]
```diff
@@ -1267,20 +1308,40 @@ void test_MQTTAgent_CommandLoop_add_acknowledgment_failure( void )
...
-
+    /* Call API under test. */
     mqttStatus = MQTTAgent_CommandLoop( &mqttAgentContext );
 
     TEST_ASSERT_EQUAL( MQTTNoMemory, mqttStatus );
 
     /* Ensure that callback is invoked. */
     TEST_ASSERT_EQUAL( 1, commandCompleteCallbackCount );
+
+    MQTTAgentCommand_Publish_ExpectAnyArgsAndReturn( MQTTSuccess );
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1328]: Can you reset the value of `commandCompleteCallbackCount` right here? It looks like the only reason it should be 2 later is that a test was run once before.

File: source/mqtt_agent.c [hunk 6]
```diff
@@ -814,11 +845,23 @@ static void clearPendingAcknowledgments( MQTTAgentContext_t * pMqttAgentContext
     {
         if( pendingAcks[ i ].packetId != MQTT_PACKET_ID_INVALID )
         {
-            /* Receive failed to indicate network error. */
-            concludeCommand( pMqttAgentContext, pendingAcks[ i ].pOriginalCommand, MQTTRecvFailed, NULL );
+            bool clearEntry = true;
+
+            if( clearOnlySubUnsubEntries &&
```
Comment (muneebahmed10) [inline source/mqtt_agent.c:850]: Nit: Perhaps we should `assert( pendingAcks[ i ].pOriginalCommand != NULL );` before dereferencing it.

File: source/mqtt_agent.c [hunk 7]
```diff
@@ -814,11 +845,23 @@ static void clearPendingAcknowledgments( MQTTAgentContext_t * pMqttAgentContext
     {
         if( pendingAcks[ i ].packetId != MQTT_PACKET_ID_INVALID )
         {
-            /* Receive failed to indicate network error. */
-            concludeCommand( pMqttAgentContext, pendingAcks[ i ].pOriginalCommand, MQTTRecvFailed, NULL );
+            bool clearEntry = true;
+
```
Comment (aggarw13) [inline source/mqtt_agent.c:849]:
```suggestion
            bool clearEntry = true;
            
            assert( pendingAcks[ i ].pOriginalCommand != NULL );
                        
```

File: test/unit-test/mqtt_agent_utest.c (hunk 5 above)
Comment (aggarw13) [reply test/unit-test/mqtt_agent_utest.c:1335]: Sure, done.

File: test/unit-test/mqtt_agent_utest.c [hunk 8]
```diff
@@ -529,16 +529,57 @@ void test_MQTTAgent_ResumeSession_session_present_no_publish_found( void )
...
+
+    MQTT_PublishToResend_IgnoreAndReturn( MQTT_PACKET_ID_INVALID );
+
+    setupAgentContext( &mqttAgentContext );
+
+    /* Setup the pending ack list to contain operations for both SUBSCRIBE
+     * and UNSUBSCRIBE operations. */
+    mqttAgentContext.pPendingAcks[ 0 ].packetId = 1U;
+    mqttAgentContext.pPendingAcks[ 0 ].pOriginalCommand = &subscribeCommand;
+
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:564]: Should we also add a PUBLISH command and test if that does *not* get cleared?

File: test/unit-test/mqtt_agent_utest.c [hunk 9]
```diff
@@ -1239,23 +1280,24 @@ void test_MQTTAgent_CommandLoop_add_acknowledgment_success( void )
...
+ * cannot be added in the list of pending acknowledgments, either due to the array
+ * being full OR there being an existing entry for the same packet ID.
  */
 void test_MQTTAgent_CommandLoop_add_acknowledgment_failure( void )
 {
     MQTTStatus_t mqttStatus;
     MQTTAgentContext_t mqttAgentContext;
     size_t i = 0;
     Command_t commandToSend = { 0 };
+    uint16_t testPacketId = 1U;
```
Comment (muneebahmed10) [inline test/unit-test/mqtt_agent_utest.c:1293]: Could make this `const` if its value doesn't change
```suggestion
    const uint16_t testPacketId = 1U;
```

**Review by muneebahmed10 — APPROVED** [review]

File: test/unit-test/mqtt_agent_utest.c (hunk 8 above)
Comment (aggarw13) [reply test/unit-test/mqtt_agent_utest.c]: Added

**Review by n9wxu — APPROVED** [review]

**Pull request merged**